### PR TITLE
feat[lang]: add `blobhash()` builtin

### PR DESCRIPTION
**Pull request by pcaversaccio**
### What I did

Introduce the built-in function `blobhash(uint256)` for retrieving versioned hashes of BLOBs.

### How I did it

Hard skills.

### How to verify it

See https://eips.ethereum.org/EIPS/eip-4844#opcode-to-get-versioned-hashes.

### Commit message

```
This commit adds access to the `BLOBHASH` opcode via the new builtin
function `blobhash()`. The opcode was introduced in EIP-4844.

References:
https://eips.ethereum.org/EIPS/eip-4844#opcode-to-get-versioned-hashes
```

### Description for the changelog

feat: introduce the built-in function `blobhash(uint256)` for retrieving versioned hashes of BLOBs.

### Cute Animal Picture

image

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]

File: vyper/builtins/functions.py [hunk 1]
```diff
@@ -1219,6 +1221,18 @@ def build_IR(self, expr, args, kwargs, contact):
...
+class BlobHash(BuiltinFunctionT):
+    _id = "blobhash"
+    _inputs = [("index", UINT256_T)]
+    _return_type = BYTES32_T
+
+    @process_inputs
+    def build_IR(self, expr, args, kwargs, contact):
+        if not version_check(begin="cancun"):
+            raise EvmVersionException("`blobhash` is not available pre-cancun", expr)
+        return IRnode.from_list(["blockhash", args[0]], typ=BYTES32_T)
```
Comment (charles-cooper) [inline vyper/builtins/functions.py:1233]: typo

File: vyper/builtins/functions.py [hunk 2]
```diff
@@ -1219,6 +1221,18 @@ def build_IR(self, expr, args, kwargs, contact):
...
 
 
+class BlobHash(BuiltinFunctionT):
+    _id = "blobhash"
+    _inputs = [("index", UINT256_T)]
+    _return_type = BYTES32_T
+
+    @process_inputs
+    def build_IR(self, expr, args, kwargs, contact):
+        if not version_check(begin="cancun"):
```
Comment (charles-cooper) [inline vyper/builtins/functions.py:1231]: should this check move to the frontend? like if somebody requests `blobhash` but only requests `annotated_ast` should this fail? or we only fail if ir/asm/bytecode is requested

File: tests/functional/builtins/codegen/test_blobhash.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+@pytest.mark.requires_evm_version("cancun")
+@pytest.mark.parametrize("good_code", valid_list)
+def test_blobhash_success(good_code):
+    assert compiler.compile_code(good_code) is not None
+    assembly = compiler.compile_code(good_code, output_formats=["asm"])["asm"].split(" ")
+    assert "BLOBHASH" in assembly
+
+
+@pytest.mark.requires_evm_version("shanghai")
```
Comment (charles-cooper) [inline tests/functional/builtins/codegen/test_blobhash.py:47]: i don't think this is necessary, it's already covered by the above test

Comment (pcaversaccio) [reply tests/functional/builtins/codegen/test_blobhash.py:47]: hmm, we should always write positive and negative tests IMO.

File: vyper/builtins/functions.py (hunk 2 above)
Comment (pcaversaccio) [reply vyper/builtins/functions.py:1231]: Personally am not too opinionated about it, but I think failure if ir/asm/bytecode is requested seems fine. Is there maybe another similar case we can look at?

File: vyper/builtins/functions.py (hunk 1 above)
Comment (pcaversaccio) [reply vyper/builtins/functions.py]: Fixed via 69a2362a67e9eea7a3b51286ee3950eea7bda255.

File: tests/functional/builtins/codegen/test_blobhash.py (hunk 3 above)
Comment (charles-cooper) [reply tests/functional/builtins/codegen/test_blobhash.py:47]: check what `requires_evm_version` does, it already ensures that EvmVersionException is thrown if version<cancun.

File: vyper/builtins/functions.py (hunk 2 above)
Comment (charles-cooper) [reply vyper/builtins/functions.py:1231]: hmm mostly we reject versioned features during codegen, although `transient` is rejected in the frontend

File: tests/functional/builtins/codegen/test_blobhash.py (hunk 3 above)
Comment (pcaversaccio) [reply tests/functional/builtins/codegen/test_blobhash.py]: removed this test via ab08210aa9edb00c2841b6f9a67a726fae4dffba

File: vyper/builtins/functions.py (hunk 2 above)
Comment (pcaversaccio) [reply vyper/builtins/functions.py:1231]: the new block property `block.blobbasefee` is currently also planned to reject at codegen time: https://github.com/vyperlang/vyper/pull/3945/files#diff-452732981c7a1a928e9f249c7b14c723695de3d6635568ccc8899b6cab2fc54bR292-R294. We should keep this consistent at least.

**Comment by charles-cooper** [comment]
we should ideally have functional tests, like check that blobhash actually matches what is in the evm

File: tests/evm_backends/pyevm_env.py [hunk 4]
```diff
@@ -4,16 +4,16 @@
 
 import rlp
 from cached_property import cached_property
-from eth.abc import ChainAPI, ComputationAPI
+from eth.abc import ChainAPI, ComputationAPI, VirtualMachineAPI
```
Comment (charles-cooper) [inline tests/evm_backends/pyevm_env.py:7]: this is a nit but i'm not a huge fan of importing things that are only used in type signatures, i am starting to prefer guarding these with `if TYPE_CHECKING`

**Pull request merged**